### ISSUE TITLE
flushAndWait in PutMarkLogicRecord causes significant delays

### ISSUE THREAD

**Issue by rjrudin** [marklogic/nifi#43]
I'm ingesting CSV data and using GetFile -> SplitText -> PutMarkLogicRecord with a CSVReader and a JSONRecordSetWriter, but the ingest rate is less than 400 docs/s, commpared to over 2k for mlcp with the same data in the same environment. I modified PMLR locally and added some logging to see where the delay is - this shows two batches, each with 100 CSV lines in them:

```
2019-01-15 11:27:38,635 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Starting: 0
2019-01-15 11:27:38,641 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Flushing: 6
2019-01-15 11:27:38,641 INFO [Timer-Driven Process Thread-7] c.m.c.datamovement.impl.WriteBatcherImpl flushing 0 queued docs
2019-01-15 11:27:39,420 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Flushed: 785
2019-01-15 11:27:39,420 INFO [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Inserted 100 records into MarkLogic
2019-01-15 11:27:39,424 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Starting: 0
2019-01-15 11:27:39,424 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] UFFM size: 0: 0
2019-01-15 11:27:39,429 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Flushing: 5
2019-01-15 11:27:39,429 INFO [Timer-Driven Process Thread-7] c.m.c.datamovement.impl.WriteBatcherImpl flushing 0 queued docs
2019-01-15 11:27:40,202 WARN [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Flushed: 778
2019-01-15 11:27:40,203 INFO [Timer-Driven Process Thread-7] o.a.n.m.processor.PutMarkLogicRecord PutMarkLogicRecord[id=4f6293dd-0168-1000-a7eb-1c32960dacff] Inserted 100 records into MarkLogic
```

You can see how the flushAndWait call is taking several hundred milliseconds each time. Changing this to flushAsync or removing it speeds things up a great deal, though this leads to a null pointer with uriFlowFileMap. But I think that can be avoided too - I'm putting together a PR for this, just wanted to track this.

**Comment by rjrudin** [comment]
The benefit of calling flushAndWait here is that the onBatchSuccess listener is guaranteed to finish first, which then allows PMLR to safely remove the corresponding entry from the uriFlowFileMap (should really be named uuidFlowFileMap, that's a separate matter). And that means that while the success listener is running, it can always find the appropriate FlowFile object for the given WriteEvent, because the FlowFile hasn't been removed from the map yet.

This all falls apart though when flushAndWait is removed. Now if there are e.g. 100 lines associated with a FlowFile, those are all added to the WriteBatcher, during which time other batches are being added too (assuming that the PMLR has more than one concurrent task, which seems worth doing to speed up the reading of FlowFiles). When a WriteBatch is created, it will likely contain documents from multiple FlowFiles. And now in the success listener, there's no way to know (yet) if all of the documents associated with a FlowFile have been written to ML yet. 

I think to do that, we'll need to keep track of how many WriteEvents were generated for a FlowFile. And then in routeDocumentToRelationship, after processing a WriteEvent, we'll see if we've seen that number of WriteEvents for the FlowFile that it's associated with. If so, then it can safely be removed from the uuidFlowFileMap.

**Comment by rjrudin** [comment]
@ehennum @ryanjdew I mis-diagnosed this - I was looking at the "Inserted N records" logging to determine when a job had finished, but that's logged only when the WriteEvents are added to a batch. When I added logging to the batch success listener, I noticed that there's little difference in performance between flushAndWait and removing it or changing it to flushAsync.

The culprit is the DHF REST transform. When I removed that, performance went from about 300 docs/s to 2k/s, similar to mlcp. Sigh. 

I'm going to submit a PR to improve the logging a bit so that it's more apparent how long it's taking to write batches.
